### PR TITLE
Remove light merging from LocalChunkProvider and refactor completeUpdate

### PR DESCRIPTION
**Pull request by oniatus**
This PR prepares an overhaul of the ChunkProviding system which is coupled to a different othere concepts like light merging, chunk loading and network transfer of chunks.

Changing any behavior of the system is out of scope of this PR and will be done in a follow up.
Therefore the goal here is just to put the system under test and do small refactorings to make the current API more expressive, even if it means to keep bugs or inintuitive concepts for the moment.

### CODE REVIEW

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine-tests/src/test/java/org/terasology/world/chunks/localChunkProvider/LocalChunkProviderTest.java [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+
+    @Test
+    public void testCompleteUpdateHandlesFinalizedChunkIfReady() throws Exception {
+        final Chunk chunk = mockChunkAt(0, 0, 0);
+        final ReadyChunkInfo readyChunkInfo = ReadyChunkInfo.createForNewChunk(chunk, new TShortObjectHashMap<>(), Collections.emptyList());
+        when(chunkFinalizer.completeFinalization()).thenReturn(readyChunkInfo);
+
+        chunkProvider.completeUpdate();
+
+        final InOrder inOrder = inOrder(worldEntity);
```
Comment (flo) [inline engine-tests/src/test/java/org/terasology/world/chunks/localChunkProvider/LocalChunkProviderTest.java]: I think it is confusing when a staticly imported method and a variable have the same name.

File: engine-tests/src/test/java/org/terasology/world/chunks/localChunkProvider/LocalChunkProviderTest.java [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+        final ReadyChunkInfo readyChunkInfo = ReadyChunkInfo.createForRestoredChunk(chunk, blockPositionMappings, mock(ChunkStore.class), Collections.emptyList());
+        when(chunkFinalizer.completeFinalization()).thenReturn(readyChunkInfo);
+
+        chunkProvider.completeUpdate();
+
+        final ArgumentCaptor<Event> eventArgumentCaptor = ArgumentCaptor.forClass(Event.class);
+        verify(blockEntity, atLeastOnce()).send(eventArgumentCaptor.capture());
+        final Event event = eventArgumentCaptor.getAllValues().get(0);
+        assertThat(event, instanceOf(OnAddedBlocks.class));
+        assertThat(((OnAddedBlocks) event).getBlockPositions(), hasItem(new Vector3i(1, 2, 3)));
```
Comment (flo) [inline engine-tests/src/test/java/org/terasology/world/chunks/localChunkProvider/LocalChunkProviderTest.java]: Maybe a TODO comment here would be good for indicating that it is a bit strange to have an OnAddedBlocks event for a restored chunk.

File: engine/src/main/java/org/terasology/world/chunks/localChunkProvider/ChunkCache.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ */
+package org.terasology.world.chunks.localChunkProvider;
+
+import org.terasology.math.geom.Vector3i;
+import org.terasology.world.chunks.Chunk;
+
+import java.util.Collection;
+import java.util.Iterator;
+
+interface ChunkCache {
```
Comment (flo) [inline engine/src/main/java/org/terasology/world/chunks/localChunkProvider/ChunkCache.java]: I think interfaces and classes should have a small javadoc that informs about purpose.

File: engine/src/main/java/org/terasology/world/chunks/localChunkProvider/ChunkFinalizer.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.world.chunks.localChunkProvider;
+
+import org.terasology.world.chunks.Chunk;
+import org.terasology.world.chunks.internal.GeneratingChunkProvider;
+import org.terasology.world.chunks.internal.ReadyChunkInfo;
+
+public interface ChunkFinalizer {
```
Comment (flo) [inline engine/src/main/java/org/terasology/world/chunks/localChunkProvider/ChunkFinalizer.java]: I think interfaces and classes should have a small javadoc that informs about purpose.

File: engine-tests/src/test/java/org/terasology/world/chunks/localChunkProvider/LocalChunkProviderTest.java [hunk 5]
```diff
@@ -0,0 +1,314 @@
+/*
```
Comment (Cervator) [inline engine-tests/src/test/java/org/terasology/world/chunks/localChunkProvider/LocalChunkProviderTest.java]: 2018 :D 

Looking forward to running tests when able

File: engine/src/main/java/org/terasology/world/chunks/localChunkProvider/ChunkFinalizer.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    void initialize(GeneratingChunkProvider generatingChunkProvider);
+
+    ReadyChunkInfo completeFinalization();
+
+    void beginFinalization(Chunk chunk, ReadyChunkInfo readyChunkInfo);
+
+    void restart();
+
+    void shutdown();
+}
```
Comment (Cervator) [inline engine/src/main/java/org/terasology/world/chunks/localChunkProvider/ChunkFinalizer.java]: Totally missing a finalize method that sets egg to null, right? ;-)

Comment (oniatus) [reply engine/src/main/java/org/terasology/world/chunks/localChunkProvider/ChunkFinalizer.java]: 😅  "At first there was the null"

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by flo** [comment]
Merged! Thanks for good work!

**Pull request merged**